### ISSUE TITLE
Support aarch64

### ISSUE THREAD

**Issue by davidlattimore** [davidlattimore/wild#43]
The main code in the linker that would be different will be the code in `relaxation.rs`. Any code that directly references x86-64 relocation types will also need to change, although outside of `relaxation.rs` there's not too much of that.

linker-diff will need some changes. A disassembler for aarch64 will need to be selected and used in `asm_diff.rs`.

This will likely be the first port of Wild. Some care will be needed to make sure that performance doesn't degrade. The simplest option would be to switch between different bits of code (e.g. different relaxation schemes) at each call site via either a dyn object, or more likely an enum. Using a dyn object would prevent inlining, so might affect performance. Using an enum wouldn't prevent inlining, so might be faster. Even with an enum, we'd still be branching. It's likely that an enum will be sufficient and that there will be no measurable slow-down. However if that's not the case, then 3rd option would be make the code generic over a parameter `P: Platform`. Switching between different implementations can then be done at some outer, less-hot point in the code.

I'd be very tempted to just jump straight to the generic-over-platform option, but don't want to preclude the other options.

One advantage of the code being generic over platform is that the relaxation code can use platform-specific types. e.g. maybe the enum `RelaxationKind` might need to be different depending on the platform. However maybe it'd be better to move the little bits of code that use a type like `RelaxationKind` into `relaxation.rs` and make `RelaxationKind` private, then the type can become an internal implementation detail of the platform and doesn't need to be exposed at all.

**Comment by andrewdavidmackenzie** [comment]
This refers to supporting aarch64 targets for the code generated, not (necessarily) running on aarch64, right?

That (for running on mac apple silicon, Raspberry PI, or other ARM PCs that are emerging) would need a separate issue I assume.

**Comment by davidlattimore** [comment]
> This refers to supporting aarch64 targets for the code generated, not (necessarily) running on aarch64, right?

Yep, it's about the code generated. I wouldn't expect any substantial work to be required to actually run the linker on the other targets. I'd certainly expect that it would already run fine on a Raspberry Pi (if you gave it X86_64 objects to link).

**Comment by davidlattimore** [comment]
I should mention that I submitted a [change yesterday](https://github.com/davidlattimore/wild/pull/191) that added an `Arch` trait. This was done to make porting to ARM and other targets easier. I figured I'd add the trait because it's kind of an invasive change. I'm not currently working on actually porting to any other architectures, so you or anyone else should feel free to do so. Also, don't feel that the trait is set in stone. It can definitely be changed as needed.

**Comment by marxin** [comment]
I am kind of thinking about doing the Arm64 port :crossed_fingers: , but I would like to get more info before I get to it:

> linker-diff will need some changes. A disassembler for aarch64 will need to be selected and used in `asm_diff.rs`.

Apparently, there are not many alternatives for Arm64; the only Rust I found is https://crates.io/crates/disarm64, can you evaluate if it's going to provide what we need in asm-diff tool?

Speaking about potential differences: Arm64 comes with its own set of relocation types, do you have any estimation of how's that different from x86_64? What about the addressing modes? Will we need a special handling of the TLS relocations?

Thanks for the replies, and based on the reply, I might commit to work on the PR.

**Comment by andrewdavidmackenzie** [comment]
Out of interest, what type or arm64/aarch64 targets were you thinking about targetting?
I use Apple Silicon Mac, and Raspberry Pis that are aarch64, so was thinking about what devices it could be supported/tested on?
(FYI: GH Actions will give you a aarch64 machine for macos-15)

**Comment by marxin** [comment]
To be honest, I am pretty new to the Arm64 platform. Definitely, both of the aforementioned CPUs should be implementation targets as both are ARMv8 (or newer). Note the Apple silicon run needs to happen on a Linux system (https://asahilinux.org/fedora/, Asahi Linux or something similar).

**Comment by davidlattimore** [comment]
> I am kind of thinking about doing the Arm64 port

Great!

> Apparently, there are not many alternatives for Arm64; the only Rust I found is https://crates.io/crates/disarm64, can you evaluate if it's going to provide what we need in asm-diff tool?

I'm sure it would be possible. Whether it would be straightforward, I'm unsure. From a quick look at the API, one thing that stands out is that I'm not seeing any way to get the operands for an instruction without first matching on the specific operation. However it might be possible, by taking the original relocation into account to come up with a different way of diffing that would mean that wouldn't matter. I'm not entirely satisfied with the way the x86-64 diffing works. The current design originated from me trying to support diffing the two executables without knowledge of the original object files. In retrospect, I think that was a mistake and we should probably just require the original object files and make full use of the information in them... I started writing that up here, but then decided it should be a separate issue, so I filed #213.

> Speaking about potential differences: Arm64 comes with its own set of relocation types, do you have any estimation of how's that different from x86_64?

Oh right, I forgot to take relocations into account when adding the `Arch` trait. So I guess that'd be the first step - to extend the trait in that regard. I haven't yet looked through the different relocation types for ARM. Wild currently converts the x86-64 relocation types into its own internal enum `RelocationKind`. Assuming there's lots of overlap between ARM and x86-64 relocation types, that might be helpful, since we'd be able to reuse the existing code for common relocation types.

> What about the addressing modes?

I haven't read it in any detail yet, but Maskray (lld maintainer) has a blog post about aarch64 from a linker perspective. Still uses the GOT (global offset table).

Range thunks look like something we might need to handle at some point, but they're only needed when jumping more than 128MiB, so we could, if we wanted skip them in the initial implementation.

> Will we need a special handling of the TLS relocations?

It does look like there are some differences. At least it sounds like TLS global and local dynamic aren't used. They're one of the more complicated things in x86-64, so that's nice. I get the impression that what's used instead sounds a little bit like the local-exec model for x86-64, although I could be mistaken.

**Comment by davidlattimore** [comment]
Just to clarify, this issue isn't about porting to macOS, just about porting to aarch64. Porting to macOS would require writing the executable format used on macOS - [Mach-O](https://en.wikipedia.org/wiki/Mach-O). Once we support aarch64, we could potentially run on Mac hardware, but only if it's running Linux.

**Comment by andrewdavidmackenzie** [comment]
Understood.

That makes we wonder about the actual target for this:
- cross-compiling/linking from a linux box running x86?
- native on some linux box running arm64?
  - there aren't too many of them (yet), except Pi...and they are a bit under-powered

**Comment by davidlattimore** [comment]
Asahi Linux and Raspberry Pi are two popular platforms, but there are also a small, but increasing number of ARM based laptops.

It'd also be good to eventually port to MacOS and supporting aarch64 is a step that would make that easier.

**Comment by marxin** [comment]
All right, I am going to start working on the Arm64 port, it's tempting to me :)

**Comment by marxin** [comment]
Have @davidlattimore a couple more question regarding the AArch64 relocations:

- The architecture has a couple of `G(GDAT(S))` relocations (which is an address of the GOT entry for symbol `S`). Am I correct that `RelocationKind` does not define the situation yet? Seems quite simple to add.

- Various relocations expect a certain range of values that can be verified (e.g. `Check that -231 <= X < 232`). What's your take on this? Are there any similar requirements when it comes to x86_64? Should I implement the ragen checking, or is it something we can rely on?

- There's a special operand adjustment present in various relocations:

```
Page(expr) is the page address of the expression expr, defined as (expr & ~0xFFF). (This applies even if the machine page size supported by the platform has a different value.)
```

An example usage can be something like `Page(S+A)-Page(P)`. To support these, I don't want to add a special `RelocationKind` enum variant, but rather I would come up with an optional operand modifier that will be applied in the relocation value calculation. So for something like:

```
        RelocationKind::Relative => resolution
            .value_with_addend(
                addend,
                symbol_index,
                object_layout,
                &layout.merged_strings,
                &layout.merged_string_start_addresses,
            )?
            .wrapping_sub(place),
...
        RelocationKind::Relative => adjust_s_plus_a(resolution
            .value_with_addend(
                addend,
                symbol_index,
                object_layout,
                &layout.merged_strings,
                &layout.merged_string_start_addresses,
            )?)
            .wrapping_sub(adjust_place(place)),
```

Do you support this approach? Or should I choose a different implementation path?

**Comment by davidlattimore** [comment]
> * The architecture has a couple of `G(GDAT(S))` relocations (which is an address of the GOT entry for symbol `S`). Am I correct that `RelocationKind` does not define the situation yet? Seems quite simple to add.

Yep, I don't think that's something I've needed. I'm guessing this is only used for non-relocatable binaries? I can't see how it would work for a position-independent executable.

> * Various relocations expect a certain range of values that can be verified (e.g. `Check that -231 <= X < 232`). What's your take on this? Are there any similar requirements when it comes to x86_64? Should I implement the ragen checking, or is it something we can rely on?

If there's anything like that for x86_64, then it's not currently being checked. I assume if the check fails, that we fail the link?


> An example usage can be something like `Page(S+A)-Page(P)`. To support these, I don't want to add a special `RelocationKind` enum variant, but rather I would come up with an optional operand modifier that will be applied in the relocation value calculation.

I assume this is used in combination with a few different relocation kinds? i.e that if you'd need to add a few new `RelocationKind`s if you went that route. If that's the case, then what you propose sounds reasonable.

**Comment by marxin** [comment]
> Yep, I don't think that's something I've needed. I'm guessing this is only used for non-relocatable binaries?

I guess so. It's definitely used in the `weaks-fns.c` integration test-case. The implementation is present in my PR now.

> I assume if the check fails, that we fail the link?

Yep, it's probably something we might consider adding in the future.

I've decided to implement it with `PageMask` (and `PageMaskValue`), the implementation seems to me quite straightforward and pretty fast as masking is done unconditionally (by default u64::MAX value is used for masking).

**Comment by marxin** [comment]
After spending some time with debugging I discovered one unpleasant thing related to the AArch64 relocations: the bit range extracting relocations speak about what bits of the calculated value are used. Though what's uncovered, where exactly are these bits placed. Apparently, each relocation is specific to an instruction, and the encoding of the immediate differs: e.g. for [ADRP](https://developer.arm.com/documentation/ddi0602/2024-12/Base-Instructions/ADRP--Form-PC-relative-address-to-4KB-page-?lang=en), the immediate is encoded in 2 chunks ([30:29], [23:5]) and e.g. [MOVZ](https://developer.arm.com/documentation/ddi0602/2024-12/Base-Instructions/MOVZ--Move-wide-with-zero-?lang=en) is encoded as one block [5:20]. Plus, there are also relocations that select between opcodes (MOVN, MOVZ) of an instruction based on the calculated relocation value. 

That said, I tend to come up with a custom handling of the relocations for AArch64, where `apply_relocation` will call a AArch64-specific function with all the components (S, A, GOT, ...) and the function will be responsible for the encoding and storing of the final value.

What do you think @davidlattimore?

**Comment by davidlattimore** [comment]
So this means that `RelocationKind` doesn't really fit ARM and the idea is to skip that abstraction and have separate code for handling ARM relocations? If ARM relocations are different enough, then that seems reasonable to me.

**Comment by marxin** [comment]
Well, my first attempt would likely be extending `RelocationSize::BitRange`, where I'll add an enum for AArch64 instructions, where the option values will drive how a computed value is stored to the underlying buffer. As I'm thinking about it, it should be pretty straightforward code.

**Comment by marxin** [comment]
> Range thunks look like something we might need to handle at some point, but they're only needed when jumping more than 128MiB, so we could, if we wanted skip them in the initial implementation.

#1763
